src/simulators/statevector/qubitvector_thrust.hpp

```diff
@@ -0,0 +1,3146 @@
+}
+
+
+/*******************************************************************************
+ *
+ * APPLY OPTIMIZED GATES
+ *
+ ******************************************************************************/
+
+//------------------------------------------------------------------------------
+// Multi-controlled gates
+//------------------------------------------------------------------------------
+
+template <typename data_t>
+struct CX_lambda : public unary_function<uint_t,void>
+{
+	thrust::complex<data_t>* pVec;
+	uint_t add;
+	uint_t mask;
+
+	CX_lambda(thrust::complex<data_t>* pV,uint_t* qubits,int nqubits)
+	{
+		int i;
+		pVec = pV;
+
+		add = 1ull << qubits[nqubits-1];
+		mask = 0;
+		for(i=0;i<nqubits-1;i++){
+			mask |= (1ull << qubits[i]);
+		}
+	}
+
+	__host__ __device__ void operator()(const uint_t &i) const
+	{
+		uint_t ip;
+		thrust::complex<data_t> q0,q1;
+
+		if((i & mask) == mask){
+			ip = i ^ add;
+			if(i < ip){
+				q0 = pVec[i];
+				q1 = pVec[ip];
+
+				pVec[i ] = q1;
+				pVec[ip] = q0;
+			}
+		}
+	}
+};
+
+template <typename data_t>
+void QubitVectorThrust<data_t>::apply_mcx(const reg_t &qubits) {
+  // Calculate the permutation positions for the last qubit.
+  const size_t N = qubits.size();
+
+#ifdef QASM_TIMING
+		TimeStart(QS_GATE_CX);
+#endif
+
+	apply_lambda(CX_lambda<data_t>((thrust::complex<data_t>*)&data_[0],(uint_t*)&qubits[0], N), data_size_);
+
+#ifdef QASM_TIMING
+		TimeEnd(QS_GATE_CX);
+#endif
+}
+
+
+template <typename data_t>
+struct CY_lambda : public unary_function<uint_t,void>
+{
+	thrust::complex<data_t>* pVec;
+	uint_t add;
+	uint_t mask;
+
+	CY_lambda(thrust::complex<data_t>* pV,uint_t* qubits,int nqubits)
+	{
+		int i;
+		pVec = pV;
+
+		add = 1ull << qubits[nqubits-1];
+		mask = 0;
+		for(i=0;i<nqubits-1;i++){
+			mask |= (1ull << qubits[i]);
+		}
+	}
+
+	__host__ __device__ void operator()(const uint_t &i) const
+	{
+		uint_t ip;
+		thrust::complex<data_t> q0,q1;
+
+		if((i & mask) == mask){
+			ip = i ^ add;
+			if(i < ip){
+				q0 = pVec[i];
+				q1 = pVec[ip];
+
+				pVec[i ] = thrust::complex<data_t>(q1.imag(),-q1.real());
+				pVec[ip] = thrust::complex<data_t>(-q0.imag(),q0.real());
+			}
+		}
+	}
+};
+
+template <typename data_t>
+void QubitVectorThrust<data_t>::apply_mcy(const reg_t &qubits) {
+  // Calculate the permutation positions for the last qubit.
+  const size_t N = qubits.size();
+//		printf("  Y\n ");
+
+	apply_lambda(CY_lambda<data_t>((thrust::complex<data_t>*)&data_[0],(uint_t*)&qubits[0], N), data_size_);
+}
+
+template <typename data_t>
+void QubitVectorThrust<data_t>::apply_mcswap(const reg_t &qubits) {
+  // Calculate the swap positions for the last two qubits.
+  // If N = 2 this is just a regular SWAP gate rather than a controlled-SWAP gate.
+	const size_t N = qubits.size();
+
+	uint_t size = data_size_ >> N;
+	uint_t pos[2];
+
+	allocate_buffers(N);
+
+	pos[0] = (1ull << 
```

Comment:
These kinds of print statements will break being able to return a valid result object if it prints to `std::out`, it would be better if they could be added to result metadata somehow.